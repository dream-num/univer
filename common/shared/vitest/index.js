/**
 * Copyright 2023-present DreamNum Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

const { defineConfig, mergeConfig } = require('vitest/config');

function createConfig(options) {
    return defineConfig(mergeConfig({
        test: {
            css: {
                modules: {
                    classNameStrategy: 'non-scoped',
                },
            },
            environment: 'happy-dom',
            coverage: {
                reporter: ['html', 'json'],
                provider: 'custom',
                customProviderModule: require.resolve('@vitest/coverage-istanbul'),
                exclude: [
                    'coverage/**',
                    'dist/**',
                    '**/[.]**',
                    'packages/*/test?(s)/**',
                    '**/*.d.ts',
                    '**/virtual:*',
                    '**/__x00__*',
                    '**/\x00*',
                    'cypress/**',
                    'test?(s)/**',
                    'test?(-*).?(c|m)[jt]s?(x)',
                    '**/*{.,-}{test,spec}?(-d).?(c|m)[jt]s?(x)',
                    '**/__test?(s)__/**',
                    '**/{karma,rollup,webpack,vite,vitest,jest,ava,babel,nyc,cypress,tsup,build}.config.*',
                    '**/vitest.{workspace,projects}.[jt]s?(on)',
                    '**/.{eslint,mocha,prettier}rc.{?(c|m)js,yml}',
                    'lib/**',
                    'src/locale/**',
                    '**/*.stories.tsx',
                    '**/__testing__/**',
                ],
            },
        },
    }, options));
}

module.exports = createConfig;
