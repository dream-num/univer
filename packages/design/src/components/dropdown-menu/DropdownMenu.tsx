/**
 * Copyright 2023-present DreamNum Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import type { Content } from '@radix-ui/react-popover';
import type { ComponentProps, ReactNode } from 'react';
import { useState } from 'react';
import {
    DropdownMenuContent,
    DropdownMenuItem,
    DropdownMenuPrimitive,
    DropdownMenuRadioGroup,
    DropdownMenuRadioItem,
    DropdownMenuSeparator,
    DropdownMenuTrigger,
} from './DropdownMenuPrimitive';
// import { DropdownMenuCheckboxItem, DropdownMenuContent, DropdownMenuGroup, DropdownMenuItem, DropdownMenuLabel, DropdownMenuPortal, DropdownMenuPrimitive, DropdownMenuRadioGroup, DropdownMenuRadioItem, DropdownMenuSeparator, DropdownMenuShortcut, DropdownMenuSub, DropdownMenuSubContent, DropdownMenuSubTrigger, DropdownMenuTrigger } from './DropdownMenuPrimitive';

type DropdownMenu = {
    type: 'item';
    children: ReactNode;
    onSelect?: (item: DropdownMenu) => void;
} | {
    type?: 'separator';
} | {
    type?: 'radio';
    options: { value: string; label: string }[];
    value: string;
    onSelect?: (item: string) => void;
};

export interface IDropdownProps {
    children: ReactNode;
    /** @deprecated */
    overlay?: ReactNode;
    items: DropdownMenu[];
    disabled?: boolean;
    open?: boolean;
    onOpenChange?: (open: boolean) => void;
}

export function DropdownMenu(props: IDropdownProps & ComponentProps<typeof Content>) {
    const {
        children,
        overlay,
        items,
        disabled,
        open: controlledOpen,
        onOpenChange: controlledOnOpenChange,
        ...restProps
    } = props;

    const [uncontrolledOpen, setUncontrolledOpen] = useState(false);

    const isControlled = controlledOpen !== undefined;
    const open = isControlled ? controlledOpen : uncontrolledOpen;

    function handleChangeOpen(newOpen: boolean) {
        if (disabled) return;

        if (!isControlled) {
            setUncontrolledOpen(newOpen);
        }

        controlledOnOpenChange?.(newOpen);
    }

    // const [showStatusBar, setShowStatusBar] = useState<boolean>(true);
    // const [showActivityBar, setShowActivityBar] = useState<boolean>(false);
    // const [showPanel, setShowPanel] = useState<boolean>(false);
    // const [position, setPosition] = useState('bottom');

    function renderMenuItem(item: DropdownMenu, index: number) {
        // const { type, children, icon, checked, hidden, onSelect } = item;
        const { type } = item;

        if (type === 'separator') {
            return <DropdownMenuSeparator key={index} />;
        } else if (type === 'radio') {
            return (
                <DropdownMenuRadioGroup
                    key={index}
                    value={item.value}
                    onValueChange={item.onSelect}
                >
                    {item.options.map((option) => (
                        <DropdownMenuRadioItem
                            key={option.value}
                            value={option.value}
                        >
                            {option.label}
                        </DropdownMenuRadioItem>
                    ))}
                </DropdownMenuRadioGroup>
            );
        } else if (type === 'item') {
            return (
                <DropdownMenuItem
                    key={index}
                    onSelect={() => {
                        item.onSelect?.(item);
                    }}
                >
                    {item.children}
                </DropdownMenuItem>
            );
        }
    }

    return (
        <DropdownMenuPrimitive open={open} onOpenChange={handleChangeOpen}>
            <DropdownMenuTrigger asChild>
                {children}
            </DropdownMenuTrigger>
            <DropdownMenuContent {...restProps}>

                {overlay}
                {items.map((item, index) => renderMenuItem(item, index))}
                {/*
                <DropdownMenuLabel>Appearance</DropdownMenuLabel>
                <DropdownMenuSeparator />
                <DropdownMenuCheckboxItem
                    checked={showStatusBar}
                    onCheckedChange={setShowStatusBar}
                >
                    Status Bar
                </DropdownMenuCheckboxItem>
                <DropdownMenuCheckboxItem
                    checked={showActivityBar}
                    onCheckedChange={setShowActivityBar}
                    disabled
                >
                    Activity Bar
                </DropdownMenuCheckboxItem>
                <DropdownMenuCheckboxItem
                    checked={showPanel}
                    onCheckedChange={setShowPanel}
                >
                    Panel
                </DropdownMenuCheckboxItem>

                <DropdownMenuSeparator />

                <DropdownMenuLabel>Panel Position</DropdownMenuLabel>
                <DropdownMenuSeparator />
                <DropdownMenuRadioGroup value={position} onValueChange={setPosition}>
                    <DropdownMenuRadioItem value="top">Top</DropdownMenuRadioItem>
                    <DropdownMenuRadioItem value="bottom">Bottom</DropdownMenuRadioItem>
                    <DropdownMenuRadioItem value="right">Right</DropdownMenuRadioItem>
                </DropdownMenuRadioGroup>

                <DropdownMenuSeparator />

                <DropdownMenuLabel>My Account</DropdownMenuLabel>
                <DropdownMenuSeparator />
                <DropdownMenuGroup>
                    <DropdownMenuItem>
                        Profile
                        <DropdownMenuShortcut>⇧⌘P</DropdownMenuShortcut>
                    </DropdownMenuItem>
                    <DropdownMenuItem>
                        Billing
                        <DropdownMenuShortcut>⌘B</DropdownMenuShortcut>
                    </DropdownMenuItem>
                    <DropdownMenuItem>
                        Settings
                        <DropdownMenuShortcut>⌘S</DropdownMenuShortcut>
                    </DropdownMenuItem>
                    <DropdownMenuItem>
                        Keyboard shortcuts
                        <DropdownMenuShortcut>⌘K</DropdownMenuShortcut>
                    </DropdownMenuItem>
                </DropdownMenuGroup>
                <DropdownMenuSeparator />
                <DropdownMenuGroup>
                    <DropdownMenuItem>Team</DropdownMenuItem>
                    <DropdownMenuSub>
                        <DropdownMenuSubTrigger>Invite users</DropdownMenuSubTrigger>
                        <DropdownMenuPortal>
                            <DropdownMenuSubContent>
                                <DropdownMenuItem>Email</DropdownMenuItem>
                                <DropdownMenuItem>Message</DropdownMenuItem>
                                <DropdownMenuSeparator />
                                <DropdownMenuItem>More...</DropdownMenuItem>
                            </DropdownMenuSubContent>
                        </DropdownMenuPortal>
                    </DropdownMenuSub>
                    <DropdownMenuItem>
                        New Team
                        <DropdownMenuShortcut>⌘+T</DropdownMenuShortcut>
                    </DropdownMenuItem>
                </DropdownMenuGroup>
                <DropdownMenuSeparator />
                <DropdownMenuItem>GitHub</DropdownMenuItem>
                <DropdownMenuItem>Support</DropdownMenuItem>
                <DropdownMenuItem disabled>API</DropdownMenuItem>
                <DropdownMenuSeparator />
                <DropdownMenuItem>
                    Log out
                    <DropdownMenuShortcut>⇧⌘Q</DropdownMenuShortcut>
                </DropdownMenuItem> */}
            </DropdownMenuContent>
        </DropdownMenuPrimitive>
    );
}
