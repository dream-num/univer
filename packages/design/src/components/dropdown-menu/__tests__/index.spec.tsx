/**
 * Copyright 2023-present DreamNum Co., Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { cleanup, render } from '@testing-library/react';
import { afterEach, describe, expect, it } from 'vitest';
import { DropdownMenu } from '../DropdownMenu';
import '@testing-library/jest-dom/vitest';

afterEach(cleanup);

describe('DropdownMenu', () => {
    it('should render with normal items', () => {
        const items = [
            { type: 'item' as const, children: 'Item 1' },
            { type: 'item' as const, children: 'Item 2', disabled: true },
        ];
        const { container } = render(
            <DropdownMenu items={items}>
                <button type="button">Trigger</button>
            </DropdownMenu>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with separator', () => {
        const items = [
            { type: 'item' as const, children: 'Item 1' },
            { type: 'separator' as const },
            { type: 'item' as const, children: 'Item 2' },
        ];
        const { container } = render(
            <DropdownMenu items={items}>
                <button type="button">Trigger</button>
            </DropdownMenu>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with subItem', () => {
        const items = [
            {
                type: 'subItem' as const,
                children: 'More',
                options: [
                    { type: 'item' as const, children: 'Sub 1' },
                    { type: 'item' as const, children: 'Sub 2' },
                ],
            },
        ];
        const { container } = render(
            <DropdownMenu items={items}>
                <button type="button">Trigger</button>
            </DropdownMenu>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with radio group', () => {
        const items = [
            {
                type: 'radio' as const,
                value: 'a',
                options: [
                    { label: 'A', value: 'a' },
                    { label: 'B', value: 'b' },
                ],
            },
        ];
        const { container } = render(
            <DropdownMenu items={items}>
                <button type="button">Trigger</button>
            </DropdownMenu>
        );
        expect(container).toMatchSnapshot();
    });

    it('should render with checkbox', () => {
        const items = [
            {
                type: 'checkbox' as const,
                value: 'c1',
                label: 'Check 1',
                checked: true,
            },
            {
                type: 'checkbox' as const,
                value: 'c2',
                label: 'Check 2',
                checked: false,
            },
        ];
        const { container } = render(
            <DropdownMenu items={items}>
                <button type="button">Trigger</button>
            </DropdownMenu>
        );
        expect(container).toMatchSnapshot();
    });
});
