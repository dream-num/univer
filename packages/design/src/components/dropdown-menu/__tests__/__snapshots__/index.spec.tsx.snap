// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DropdownMenu > should render with checkbox 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    data-slot="dropdown-menu-trigger"
    data-state="closed"
    id="radix-«r8»"
    type="button"
  >
    Trigger
  </button>
</div>
`;

exports[`DropdownMenu > should render with normal items 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    data-slot="dropdown-menu-trigger"
    data-state="closed"
    id="radix-«r0»"
    type="button"
  >
    Trigger
  </button>
</div>
`;

exports[`DropdownMenu > should render with radio group 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    data-slot="dropdown-menu-trigger"
    data-state="closed"
    id="radix-«r6»"
    type="button"
  >
    Trigger
  </button>
</div>
`;

exports[`DropdownMenu > should render with separator 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    data-slot="dropdown-menu-trigger"
    data-state="closed"
    id="radix-«r2»"
    type="button"
  >
    Trigger
  </button>
</div>
`;

exports[`DropdownMenu > should render with subItem 1`] = `
<div>
  <button
    aria-expanded="false"
    aria-haspopup="menu"
    data-slot="dropdown-menu-trigger"
    data-state="closed"
    id="radix-«r4»"
    type="button"
  >
    Trigger
  </button>
</div>
`;
